src/hotspot/share/opto/node.hpp

```diff
@@ -1628,6 +1635,35 @@ class Node_List : public Node_Array {
   void dump_simple() const;
 };
 
+// Definition must appear after complete type definition of Node_List
+template <typename Callback, typename Check>
+void Node::visit_uses(Callback callback, Check is_boundary) const {
+  ResourceMark rm;
+  VectorSet visited;
+  Node_List worklist;
+
+  // The initial worklist consists of the direct uses
+  for (DUIterator_Fast kmax, k = fast_outs(kmax); k < kmax; k++) {
+    Node* out = fast_out(k);
+    if (!visited.test_set(out->_idx)) { worklist.push(out); }
```

Comment:
```suggestion
    visited.set(out->_idx);
    worklist.push(out);
```

Reply:
Please disregard this suggestion, I see now that you want to handle duplicate direct uses here.

Reply:
Yes, I moved the visited check from the beginning of the while-loop to where we push to the worklist (suggestion by Emanuel). So now, every node is added at most once to the worklist.